sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,477 @@
+"""Capability Requests management for plugins.
+
+.. versionadded:: 8.0
+
+.. important::
+
+    This is all relatively new. Its usage and documentation is for Sopel core
+    development and advanced developers. It is subject to rapid changes
+    between versions without much (or any) warning.
+
+    Do **not** build your plugin based on what is here, you do **not** need to.
+"""
+
+from __future__ import annotations
+
+import logging
+from typing import (
+    Dict,
+    FrozenSet,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    TYPE_CHECKING,
+    Union,
+)
+
+
+if TYPE_CHECKING:
+    from sopel.bot import Sopel, SopelWrapper
+    from sopel.plugin import capability, CapabilityNegotiation
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Manager:
+    """Manager of plugins' capability requests.
+
+    Whenever a plugin declares a capability request (through
+    :class:`sopel.plugin.capability`), the bot will register the request with
+    this manager.
+
+    The bot is responsible to call the manager at the appropriate time, and the
+    manager will store requests' state and handle acknowledgement, by following
+    this workflow:
+
+    1. the bot will register plugins' requests with the :meth:`register` method
+    2. the bot will request the list of available capabilities with the
+       ``CAP LS`` subcommand
+    3. upon receiving the list, it'll use the :meth:`request_available` method
+       to let the manager send the required ``CAP REQ`` messages.
+    4. when the server ``ACK`` a request, the bot will call the
+       :meth:`acknowledge` method; when the server ``NAK`` a request, the bot
+       will call the :meth:`deny` method
+    5. once all requests are handled (either directly or after calling the
+       :meth:`resume` method), the bot will send the ``CAP END`` message to end
+       capability negotiation
+
```

Comment:
Does this blank line affect rendering? It seems like it shouldn't be necessary.

```suggestion
```

Reply:
It doesn't. I think I just like the extra space after a list.